kotlin/codegen/src/main/java/com/squareup/moshi/kotlin/codegen/Metadata.kt

```diff
@@ -0,0 +1,350 @@
+          }
+          val typeVariableName = if (upperBoundList.isEmpty()) {
+            TypeVariableName(
+                name = name,
+                variance = finalVariance
+            )
+          } else {
+            TypeVariableName(
+                name = name,
+                bounds = *(upperBoundList.toTypedArray()),
+                variance = finalVariance
+            )
+          }
+          typeParameters[id] = typeVariableName
+              .reified(Flag.TypeParameter.IS_REIFIED(flags))
+        }
+      }
+    }
+
+    override fun visitCompanionObject(name: String) {
+      companionObjectName = name
+    }
+
+    override fun visitConstructor(flags: Flags): KmConstructorVisitor? {
+      return if (Flag.Constructor.IS_PRIMARY(flags)) {
+        object : KmConstructorVisitor() {
+          val params = mutableListOf<ParameterData>()
+          override fun visitValueParameter(flags: Flags, name: String): KmValueParameterVisitor? {
+            val parameterFlags = flags
+            return object : KmValueParameterVisitor() {
+              lateinit var type: TypeName
+              var isVarArg = false
+              var varargElementType: TypeName? = null
+              override fun visitType(flags: Flags): KmTypeVisitor? {
+                return TypeNameKmTypeVisitor(flags, typeParamResolver) {
+                  type = it
+                }
+              }
+
+              override fun visitVarargElementType(flags: Flags): KmTypeVisitor? {
+                return TypeNameKmTypeVisitor(flags, typeParamResolver) {
+                  isVarArg = true
+                  varargElementType = it
+                }
+              }
+
+              override fun visitEnd() {
+                params += ParameterData(parameterFlags, name, type, isVarArg, varargElementType)
+              }
+            }
+          }
+
+          override fun visitEnd() {
+            constructorData = ConstructorData(flags, params)
+          }
+        }
+      } else {
+        null
+      }
+    }
+
+    override fun visitSupertype(flags: Flags): KmTypeVisitor? {
+      return TypeNameKmTypeVisitor(flags, typeParamResolver) {
+        superTypes += it
+      }
+    }
+
+    override fun visitProperty(flags: Flags,
+        name: String,
+        getterFlags: Flags,
+        setterFlags: Flags): KmPropertyVisitor? {
+      return object : KmPropertyVisitor() {
+        lateinit var type: TypeName
+        override fun visitEnd() {
+          properties += PropertyData(flags, name, type)
+        }
+
+        override fun visitReturnType(flags: Flags): KmTypeVisitor? {
+          return TypeNameKmTypeVisitor(flags, typeParamResolver) {
+            type = it
+          }
+        }
+      }
+    }
+  })
+
+  return ClassData(className.replace("/", "."),
+      classFlags,
+      companionObjectName,
+      constructorData,
+      superTypes,
+      typeParameters.values.toList(),
+      properties)
+}
+
+internal data class ClassData(
```

Comment:
nit: do we have a more descriptive suffix than Data? Even Metadata seems more appropriate

Reply:
Maybe this should be called `KmClass` because it appears like that’s what we’re trying to recover from the awful visitor

Reply:
`KmClass` works for me!